storage/paths/db_index.go

```diff
@@ -0,0 +1,891 @@
+	// Find all storage info which already hold this sector + filetype
+	err := dbi.harmonyDB.Select(ctx, &rows,
+		`		SELECT DISTINCT ON (stor.storage_id)
+						  stor.storage_id,
+						  COUNT(*) OVER(PARTITION BY stor.storage_id) as count, 
+						  BOOL_OR(is_primary) OVER(PARTITION BY stor.storage_id) AS is_primary,
+						  urls,
+						  weight,
+						  can_seal,
+						  can_store,
+						  groups,
+						  allow_to,
+						  allow_types,
+						  deny_types
+						FROM sectorlocation sec
+						JOIN storagelocation stor ON sec.storage_id = stor.storage_id 
+						WHERE sec.miner_id = $1
+						  AND sec.sector_num = $2 
+						  AND sec.sector_filetype = ANY($3)  
+						ORDER BY stor.storage_id`,
+		s.Miner, s.Number, fts)
+	if err != nil {
+		return nil, xerrors.Errorf("Finding sector storage from DB fails with err: %v", err)
+	}
+
+	for _, row := range rows {
+
+		// Parse all urls
+		var urls, burls []string
+		for _, u := range splitString(row.Urls) {
+			rl, err := url.Parse(u)
+			if err != nil {
+				return nil, xerrors.Errorf("failed to parse url: %w", err)
+			}
+			rl.Path = gopath.Join(rl.Path, ft.String(), storiface.SectorName(s))
+			urls = append(urls, rl.String())
+			burls = append(burls, u)
+		}
+
+		result = append(result, storiface.SectorStorageInfo{
+			ID:         storiface.ID(row.StorageId),
+			URLs:       urls,
+			BaseURLs:   burls,
+			Weight:     row.Weight * row.Count,
+			CanSeal:    row.CanSeal,
+			CanStore:   row.CanStore,
+			Primary:    row.IsPrimary,
+			AllowTypes: splitString(row.AllowTypes),
+			DenyTypes:  splitString(row.DenyTypes),
+		})
+
+		storageWithSector[row.StorageId] = true
+
+		allowTo := splitString(row.AllowTo)
+		if allowList != nil && len(allowTo) > 0 {
+			for _, group := range allowTo {
+				allowList[group] = struct{}{}
+			}
+		} else {
+			allowList = nil // allow to any
+		}
+	}
+
+	// Find all storage paths which can hold this sector if allowFetch is true
+	if allowFetch {
+		spaceReq, err := ft.SealSpaceUse(ssize)
+		if err != nil {
+			return nil, xerrors.Errorf("estimating required space: %w", err)
+		}
+
+		// Conditions to satisfy when choosing a sector
+		// 1. CanSeal is true
+		// 2. Available >= spaceReq
+		// 3. curr_time - last_heartbeat < SkippedHeartbeatThresh
+		// 4. heartbeat_err is NULL
+		// 5. not one of the earlier picked storage ids
+		// 6. !ft.AnyAllowed(st.info.AllowTypes, st.info.DenyTypes)
+		// 7. Storage path is part of the groups which are allowed from the storage paths which already hold the sector
+
+		var rows []struct {
+			StorageId  string
+			Urls       string
+			Weight     uint64
+			CanSeal    bool
+			CanStore   bool
+			Groups     string
+			AllowTypes string
+			DenyTypes  string
+		}
+		err = dbi.harmonyDB.Select(ctx, &rows,
+			`SELECT storage_id, 
+					  	urls, 
+					  	weight,
+					  	can_seal,
+					  	can_store,
+					  	groups, 
+					  	allow_types, 
+					  	deny_types
+				FROM storagelocation 
+				WHERE can_seal=true 
```

Comment:
We shouldn't be checking for can_seal here, StorageFindSector may also be used (not entirely sure if it actually is) by sector-finalization code, which wants to move the sector into a `can_store` path (which in almost all cases won't be `can_seal`)

Reply:
The current in memory index has this constraint here: https://github.com/filecoin-project/lotus/blob/77594442471f6512d5657c5be9ca4511eea4f60b/storage/paths/index.go#L491

Let me know if you think thats a bug in the current implementation as well